### ISSUE TITLE
✨ Enable the use of an Azure api-key

### ISSUE THREAD

**Issue by jaxblack** [reworkd/AgentGPT#77]
Azure api-key is different from openai api-key. And it should be used.

![image](https://user-images.githubusercontent.com/27149144/231694140-86734635-2d90-4d3e-bcae-a052ff099a30.png)

**Comment by RepubIique** [comment]
If I am not wrong from reading this 
https://msandbu.org/openai-vs-azure-openai/

Azure OpenAI API key is a specific type of API key that allows you to access OpenAI's GPT-3 language model through the Azure Cognitive Services. This means that you can use the Azure portal to create and manage your Azure OpenAI API key, and you will be billed by Microsoft Azure for any usage of the API.

On the other hand, the OpenAI API key is a general API key that allows you to access all of the OpenAI APIs, including GPT-3, through the OpenAI API platform. This means that you can create and manage your OpenAI API key through the OpenAI website, and you will be billed directly by OpenAI for any usage of the API.

In summary, the main difference between the two API keys is the platform they are associated with and the billing method used for their usage. Azure OpenAI API key is associated with Azure Cognitive Services and is billed by Microsoft Azure, while OpenAI API key is associated with the OpenAI platform and is billed by OpenAI.

**Comment by yuanman2016** [comment]
Also need this feature, hope it will be supported soon

**Comment by itsjobangrewal** [comment]
> Hi all this is definitely a something we want to support. I can support by helping to review any prs for this 😊

me too pls, waiting for next release

**Comment by karshatmosera** [comment]
We'd like this feature as well, please.

**Comment by qianchen94** [comment]
me too

**Comment by ottodeng** [comment]
me too

**Comment by KassieGuan** [comment]
me too，please

**Comment by securigy** [comment]
I am digging into Azure Semantic Kernel and its examples... and most of them reference the following:

`Env.Var("AZURE_OPENAI_EMBEDDINGS_DEPLOYMENT_NAME"),
                        Env.Var("AZURE_OPENAI_EMBEDDINGS_ENDPOINT"),
                        Env.Var("AZURE_OPENAI_EMBEDDINGS_KEY"))`
                        
I had not idea that Embeddings require separate set of Name/Endpoint/Key...
How do I get them?

**Comment by Zizico2** [comment]
I have this working internally, at my current employer, with hardcoded `api_key`s and `model_name`s, `deployment_id`s, etc.. We're just testing it out. We broke OpenAI to get Azure OpenAI working as easily as possible. Plus we still don't have access to 3.5-16k so summaries are broken. 

I might, eventually, tidy it up and submit a PR. But it is not even close to a priority currently :/. I'm more than welcome to provide any help if someone is implementing this and gets stuck.

**Comment by awtkns** [comment]
hey this was fixed in #1165

In settings.py you can now enable azure
```
    # Azure OpenAI
    azure_openai_api_version: str = "2023-06-01-preview"
    azure_openai_api_key: str = ""
    azure_openai_api_base: str = ""
    azure_openai_deployment_name: str = ""
```

Note that agentgpt uses function calling which is not supported by azure as of yet so it will fall back to openai in some cases :)

**Comment by Zizico2** [comment]
> Note that agentgpt uses function calling which is not supported by azure as of yet so it will fall back to openai in some cases :) 

I did manage to get it working, including functions. Versions 0613 support functions ([https://techcommunity.microsoft.com/t5/azure-ai-services-blog/function-calling-is-now-available-in-azure-openai-service/ba-p/3879241](source)).
 
I think Microsoft is rolling them out slowly. What I didn't manage to get working properly were the summaries because I don't have access to the 16k model. Idk if there's any solution there.

**Comment by zenyanbo** [comment]
> hey this was fixed in #1165
> 
> In settings.py you can now enable azure
> 
> ```
>     # Azure OpenAI
>     azure_openai_api_version: str = "2023-06-01-preview"
>     azure_openai_api_key: str = ""
>     azure_openai_api_base: str = ""
>     azure_openai_deployment_name: str = ""
> ```
> 
> Note that agentgpt uses function calling which is not supported by azure as of yet so it will fall back to openai in some cases :)

hellow, even though I added those options as well.
    azure_openai_api_version: str = "2023-06-01-preview"
    azure_openai_api_key: str = "<my key>"
    azure_openai_api_base: str = "https://<my domain>.openai.azure.com/"
    azure_openai_deployment_name: str = "gpt4"
However, it always show: Authentication error: Ensure a valid API key is being used. 
The next interface also only has the gpt3.5 gpt-3.5-16k gpt-4 option in the settings. It seems agentgpt is still using the openai api. What should I do?

**Comment by zenyanbo** [comment]
> hey this was fixed in #1165
> 
> In settings.py you can now enable azure
> 
> ```
>     # Azure OpenAI
>     azure_openai_api_version: str = "2023-06-01-preview"
>     azure_openai_api_key: str = ""
>     azure_openai_api_base: str = ""
>     azure_openai_deployment_name: str = ""
> ```
> 
> Note that agentgpt uses function calling which is not supported by azure as of yet so it will fall back to openai in some cases :)

hellow, even though I added those options as well.
```
    azure_openai_api_version: str = "2023-06-01-preview"
    azure_openai_api_key: str = "<my key>"
    azure_openai_api_base: str = "https://<my domain>.openai.azure.com/"
    azure_openai_deployment_name: str = "gpt4"
```
However, it always says: Authentication error: Ensure a valid API key is being used. The next interface also only has the gpt3.5 gpt-3.5-16k gpt-4 option in the settings. It seems agentgpt is still using the openai api. What should I do?

**Comment by h3nryza** [comment]
> > I have this working internally, at my current employer, with hardcoded `api_key`s and `model_name`s, `deployment_id`s, etc.. We're just testing it out. We broke OpenAI to get Azure OpenAI working as easily as possible. Plus we still don't have access to 3.5-16k so summaries are broken.
> > I might, eventually, tidy it up and submit a PR. But it is not even close to a priority currently :/. I'm more than happy to provide any help if someone is implementing this and gets stuck.
> 
> @Zizico2 How did you get this working? I've been trying to set the env variables in /next/.env and also tried changing settings via /platform/reworkd_platform/settings.py, but haven't had any luck.

I see that they deprecated the functions calling and added it under the tools API: https://learn.microsoft.com/en-us/azure/ai-services/openai/how-to/function-calling?tabs=python.

@awtkns  maybe something to add to a new or existing feature?

![image](https://github.com/reworkd/AgentGPT/assets/20166417/f51924bf-c293-44ad-b7ab-74018bd192a6)
